### PR TITLE
2-way server heartbeats (plus bonus DMAPI update)

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
The server now tracks incoming heartbeats from the server daemon (TGS). If it doesn't receive one in 2 minutes, a **very obvious** warning is thrown in chat to all online staff. Yes mentors will get this as well, and that is ok for 2 reasons.

1. If no admins are online this still needs to go out to me
2. `GLOB.admins` includes anyone with a permissions holder and this is easier. `message_admins()` didn't work with the formatting.

Might be worth TMing this just incase the tracking doesn't work properly on midnight rollovers. 

## Why It's Good For The Game
The server daemon automatically nukes the server after 5 failed heartbeats, dispatched minutely. If the server isn't receiving those, either the server daemon is dead (bad because rounds won't restart), or worse, the server daemon is trying to send heartbeats but they don't arrive or go back (very bad because current round could get nuked at a moments notice).

## Images of changes
image

## Testing
I loaded it up locally.

## Changelog
NPFC

### CODE REVIEW

**Comment by Kapu1178** [comment]
good pr let me steal

File: code/controllers/subsystem/heartbeat.dm [hunk 1]
```diff
@@ -0,0 +1,20 @@
+SUBSYSTEM_DEF(heartbeat)
+	name = "Heartbeat"
+	flags = SS_KEEP_TIMING
+	wait = 30 SECONDS
+	runlevels = RUNLEVEL_INIT | RUNLEVEL_LOBBY | RUNLEVEL_SETUP | RUNLEVEL_GAME | RUNLEVEL_POSTGAME // ALL THE THINGS
```
Comment (Kapu1178) [inline code/controllers/subsystem/heartbeat.dm:5]: why not just "ALL" to future proof against new runlevels

Comment (AffectedArc07) [reply code/controllers/subsystem/heartbeat.dm:5]: We don't have that define lmao

Comment (Kapu1178) [reply code/controllers/subsystem/heartbeat.dm:5]: gottem gg

File: code/controllers/subsystem/heartbeat.dm [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+/datum/controller/subsystem/heartbeat/Initialize()
+	// Disable if we arent running on TGS
+	if(!world.TgsAvailable())
+		flags |= SS_NO_FIRE
+
+/datum/controller/subsystem/heartbeat/fire(resumed)
+	// If the last heartbeat is 0, we never got one this round
+	if(last_heartbeat != 0 && (last_heartbeat + 2 MINUTES < REALTIMEOFDAY) && !warning_tripped)
+		to_chat(GLOB.admins, "<hr><center><span class='userdanger'><big>--- CRITICAL ---</big></span><br>The server hasn't received an uptime check from the server daemon for over 2 minutes. Inform AA ASAP.</center><hr>")
```
[the remark this answers is not in the document]
Comment (AffectedArc07) [reply code/controllers/subsystem/heartbeat.dm:19]: I could change it to that, however we (afaik) don't have downstreams + I don't plan on going anywhere anytime soon. Up to you really.

[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/controllers/subsystem/heartbeat.dm:19]: that's a fair point, its fine how it is

**Comment by AffectedArc07** [comment]
Latest DMAPI supports sending the heartbeat to the event handler which means we dont need a diff in the DMAPI

**Comment by AffectedArc07** [comment]
> Latest DMAPI supports sending the heartbeat to the event handler which means we dont need a diff in the DMAPI

So that was a lie

**Comment by AffectedArc07** [comment]
> > Latest DMAPI supports sending the heartbeat to the event handler which means we dont need a diff in the DMAPI
> 
> So that was a lie

Its been done now

**Review by S34NW — APPROVED** [review]
This has been TMd for a while any all issues addressed. Lets go.

**Pull request merged**